lib/xenops_server.ml

```diff
@@ -2319,27 +2370,19 @@ module VM = struct
     let memory_limit = List.assoc "memory_limit" cookies |> Int64.of_string in
     Debug.with_thread_associated dbg
       (fun () ->
-         let is_localhost, id = Debug.with_thread_associated dbg
+         let id, final_id = Debug.with_thread_associated dbg
              debug "VM.receive_memory";
-           let remote_instance = List.assoc "instance_id" cookies in
-           let is_localhost = instance_id = remote_instance in
+           let final_id = List.assoc "final_id" cookies in
            (* The URI is /service/xenops/memory/id *)
            let bits = Stdext.Xstringext.String.split '/' (Uri.path uri) in
            let id = bits |> List.rev |> List.hd in
-           debug "VM.receive_memory id = %s is_localhost = %b" id is_localhost;
-           is_localhost, id
+           debug "VM.receive_memory id = %s" id;
+           id, final_id
          in
          match context.transferred_fd with
          | Some transferred_fd ->
-           let op = VM_receive_memory(id, memory_limit, transferred_fd) in
-           (* If it's a localhost migration then we're already in the queue *)
-           let task =
-             if is_localhost then begin
-               immediate_operation dbg id op;
-               None
-             end else
-               Some (queue_operation dbg id op)
-           in
+           let op = VM_receive_memory(id, final_id, memory_limit, transferred_fd) in
+           let task = Some (queue_operation dbg id op) in
```

Comment:
So how do we get `VM_migrate` and `VM_receive_memory` to execute in parallel in the localhost case? Because `id`, which is `new_dest_id`, is not yet aliased at this point? But it will be once `VM_restore` is put on the queue...? I think this deserves a comment in the code to explain ;)

Reply:
https://github.com/jonludlam/xenopsd/blob/037d4657c140bd33c941de151afc3048eebba516/lib/xenops_server.ml#L1739-L1748